libvast/src/table_slice.cpp

```diff
@@ -638,42 +595,128 @@ struct row_evaluator {
+  const auto num_rows = slice.rows();
+  const auto make_path = [](const struct offset& index) noexcept {
+    auto intermediate = std::vector<int>{};
+    intermediate.reserve(index.size());
+    for (auto x : index)
+      intermediate.push_back(detail::narrow_cast<int>(x));
+    return arrow::FieldPath{std::move(intermediate)};
+  };
+  auto make_empty_result = [&] {
+    auto result = ids{};
+    result.append(false, offset + num_rows);
+    return result;
+  };
+  const auto evaluate_predicate = detail::overload{
+    [](const auto&, relational_operator, const auto&, const ids&) -> ids {
+      die("predicates must be normalized and bound for evaluation");
+    },
+    [&](const meta_extractor& lhs, relational_operator op, const data& rhs,
+        ids selection) -> ids {
+      if (!any(selection))
+        return selection;
+      if (evaluate_meta_extractor(slice, lhs, op, rhs))
+        return selection;
+      return make_empty_result();
+    },
+    [&](const data_extractor& lhs, relational_operator op, const data& rhs,
+        const ids& selection) -> ids {
+      if (!any(selection))
+        return make_empty_result();
+      const auto index
+        = caf::get<record_type>(slice.layout()).resolve_flat_index(lhs.column);
+      const auto type = caf::get<record_type>(slice.layout()).field(index).type;
+      const auto array
+        = make_path(index).Get(*to_record_batch(slice)).ValueOrDie();
+      auto result = ids{};
+      const auto rhs_internal
+        = materialize(to_internal(type, make_data_view(rhs)));
+      for (auto id : select(selection)) {
+        VAST_ASSERT(id >= offset);
+        const auto row = id - offset;
+        result.append(false, id - result.size());
+        // FIXME: This materializes unnecessarily because `evaluate_view` is
+        // broken and does not do the same thing as `evaluate`. We must align
+        // their behavior and add tests for `evaluate_view`, or, which would be
+        // even better, operate on the Arrow Arrays directly here.
+        auto lhs = materialize(
+          value_at(type, *array, detail::narrow_cast<int64_t>(row)));
+        const bool matches = evaluate(lhs, op, rhs_internal);
+        result.append(matches, 1);
+      }
+      result.append(false, offset + num_rows - result.size());
+      return result;
+    },
+  };
+  const auto evaluate_expression
+    = [&](const auto& self, const expression& expr, ids selection) -> ids {
+    const auto evaluate_expression_impl = detail::overload{
+      [&](const caf::none_t&, const ids&) {
+        return make_empty_result();
+      },
+      [&](const negation& negation, const ids& selection) {
+        return selection ^ self(self, negation.expr(), selection);
+      },
+      [&](const conjunction& conjunction, const ids& selection) {
+        auto result = selection;
+        for (const auto& connective : conjunction) {
+          if (!any(selection))
```

Comment:
can this move out of the loop? seems to not depend on it

Reply:
alternatively, I guess short-circuiting should test on `result`?

Reply:
The `result` variable is not needed at all, I think. I've changed it to just using the `selection`, and passing that in by value.